### PR TITLE
add Returns

### PR DESCRIPTION
**Pull request by jw3126**
This PR adds ~~`Always`~~ `Returns` which represents "constant" functions. So instead of writing
```julia
f = (args...; kw...) -> 42
```
you can do `f = Returns(42)`. 
Benefits are:

* Possibility to dispatch on `Returns`: 
```julia
mymap(f, x::Array) = ...
mymap(f::Returns, x::Array) = FillArrays.Fill(f.value, size(x))
```
* Easier to debug:
```
Stacktrace:
 [1] filter(Returns(42), a::UnitRange{Int64})
```
instead of
```
Stacktrace:
 [1] filter(f::var"#1#2", a::UnitRange{Int64})
```

### CODE REVIEW

**Comment by rfourquet** [comment]
`Always` sounds like a duration belonging to the `Dates` module :)

**Comment by mbauman** [comment]
Are there any cases where this is helpful in Base code itself?

**Comment by vtjnash** [comment]
Would it be too much of a pun to define this using `(c::Core.Const)(x...; y...) = c.val`? 😂

**Comment by jw3126** [comment]
> Are there any cases where this is helpful in Base code itself?

I think so. A pattern which I personally use often is `ntuple(_ -> 3, len)` and I also found this in julia. e.g. 
https://github.com/JuliaLang/julia/blob/e32e94eff3c5d0d793f51703456c499ee9958f59/test/abstractarray.jl#L302

**Comment by jw3126** [comment]
> `Always` sounds like a duration belonging to the `Dates` module :)

We can use another name of course. Names that came to my mind were `Always, Const, Constant, Return`. Looked at this naming problem in isolation `Const` would be my first choice. But as @vtjnash pointed out it collides a bit with `Core.Const`.

**Comment by JeffBezanson** [comment]
Common lisp calls this `constantly`.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2112,7 +2112,7 @@ _sub2ind_vec(i) = ()
 
 function _ind2sub(inds::Union{DimsInteger{N},Indices{N}}, ind::AbstractVector{<:Integer}) where N
     M = length(ind)
-    t = ntuple(n->similar(ind),Val(N))
+    t = ntuple(Always(similar(ind)),Val(N))
```
Comment (JeffBezanson) [inline base/abstractarray.jl:2115]: It looks like this is supposed to make a new array for each tuple element.

**Comment by MasonProtter** [comment]
Triage seems to like `Returns`

**Comment by JeffBezanson** [comment]
Suggestion from triage: call it `Returns(val)`, which should make it very clear that it's a function that ... returns the given value.

File: base/dict.jl [hunk 2]
```diff
@@ -797,7 +797,7 @@ function iterate(d::ImmutableDict{K,V}, t=d) where {K, V}
     !isdefined(t, :parent) && return nothing
     (Pair{K,V}(t.key, t.value), t.parent)
 end
-length(t::ImmutableDict) = count(x->true, t)
+length(t::ImmutableDict) = count(Always(true), t)
```
Comment (KristofferC) [inline base/dict.jl:800]: Why is this advantageous? Shouldn't it compile to the same thing?

**Comment by jw3126** [comment]
> Suggestion from triage: call it `Returns(val)`, which should make it very clear that it's a function that ... returns the given value.

Sounds good. So this means triage is in favor of adding this functionality?

File: base/operators.jl [hunk 3]
```diff
@@ -915,6 +915,40 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 """
 |>(x, f) = f(x)
 
+# always
```
Comment (JeffBezanson) [inline base/operators.jl:918]: delete

File: base/operators.jl [hunk 4]
```diff
@@ -915,6 +915,40 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
...
+"""
+struct Returns{V} <: Function
+    value::V
+    Returns{V}(value) where {V} = new{V}(value)
+    Returns(value) = new{Core.Typeof(value)}(value)
+end
+
+(obj::Returns)(args...; kw...) = obj.value
+function show(io::IO, obj::Returns)
+    show(io, Returns)
```
Comment (JeffBezanson) [inline base/operators.jl:947]:
```suggestion
    show(io, typeof(obj))
```
Of course, there are cases where the full type doesn't need to be shown, but the logic for that is quite complex so I probably wouldn't bother.

**Comment by jw3126** [comment]
I changed most of the constant lambdas in Base and test into Returns. From my point of view, this PR is ready.

**Comment by jw3126** [comment]
bump

File: NEWS.md [hunk 5]
```diff
@@ -33,6 +33,7 @@ New library functions
 * Two argument methods `findmax(f, domain)`, `argmax(f, domain)` and the corresponding `min` versions ([#27613]).
 * `isunordered(x)` returns true if `x` is value that is normally unordered, such as `NaN` or `missing`.
 * New macro `Base.@invokelatest f(args...; kwargs...)` provides a convenient way to call `Base.invokelatest(f, args...; kwargs...)` ([#37971])
+* New function/type `Returns(value)` that creates a function, which returns `value` for any arguments ([ #39794])
```
Comment (simeonschaub) [inline NEWS.md:36]:
```suggestion
* New functor `Returns(value)`, which returns `value` for any arguments ([#39794])
```

File: test/arrayops.jl [hunk 6]
```diff
@@ -1153,17 +1153,17 @@ end
...
-    n3 = mapslices(x-> fill(1,6), c, dims=[2,3])
-    n1a = mapslices(x-> fill(1,1,6), c, dims=[1,2])
-    n2a = mapslices(x-> fill(1,1,6), c, dims=[1,3])
-    n3a = mapslices(x-> fill(1,1,6), c, dims=[2,3])
+    n1 =  mapslices(Returns(fill(1,6)  ), c, dims=[1,2])
+    n2 =  mapslices(Returns(fill(1,6)  ), c, dims=[1,3])
+    n3 =  mapslices(Returns(fill(1,6)  ), c, dims=[2,3])
+    n1a = mapslices(Returns(fill(1,1,6)), c, dims=[1,2])
+    n2a = mapslices(Returns(fill(1,1,6)), c, dims=[1,3])
+    n3a = mapslices(Returns(fill(1,1,6)), c, dims=[2,3])
```
Comment (simeonschaub) [inline test/arrayops.jl:1166]: Since these are not functionally equivalent, I would prefer leaving this as-is.

Comment (jw3126) [reply test/arrayops.jl:1166]: So you say that whether only one `fill(...)` or a couple of `fill(...)` are allocated could be a detail that is relevant here?

Comment (simeonschaub) [reply test/arrayops.jl:1166]: It's probably not a big deal here, but it's often easy to miss the purpose of a test like this if you did not write it.

Comment (jw3126) [reply test/arrayops.jl]: Done!

**Review by simeonschaub — APPROVED** [review]

**Comment by jw3126** [comment]
Bunp

**Comment by rfourquet** [comment]
Was `Return` mentioned as an alternative to `Returns` ? Verbs are generally used in imperative form, like `fill`/`Fill`, `map`, `filter`, `Iterators.[tT]ake`, etc.

**Comment by jw3126** [comment]
> Was `Return` mentioned as an alternative to `Returns` ? Verbs are generally used in imperative form, like `fill`/`Fill`, `map`, `filter`, `Iterators.[tT]ake`, etc.

It was mentioned [above](https://github.com/JuliaLang/julia/pull/39794#issuecomment-784895863) but not discussed further I think.

**Comment by jw3126** [comment]
So how to move forward from here? Should this be renamed from `Returns` to `Return`?

**Comment by MasonProtter** [comment]
`Return` was discussed on the triage call, but people felt it was too different a concept from the `return` keyword.

**Comment by StefanKarpinski** [comment]
I'm amused by how every usage of this made the code longer. I suppose the advantage is that `Returns(val)` is only compiled once per argument type instead of having a different function object each type that's compiled separately.

**Comment by jw3126** [comment]
> I'm amused by how every usage of this made the code longer. I suppose the advantage is that `Returns(val)` is only compiled once per argument type instead of having a different function object each type that's compiled separately.

@StefanKarpinski in general the main selling points of `Returns` are better printing, the possibility to dispatch on it and less work for the compiler in case of repetitions. 
I think in most of the replacements in this PR, none of these are super strong arguments. It was more about consistency for me.

**Comment by mbauman** [comment]
I too was skeptical about this initially, but then immediately found myself wanting to use it for DataStructure.jl's DefaultDict, calls `Callable` default values upon access.  `DefaultDict(Returns(identity))` is more readable and obvious than `DefaultDict(x->identity)` which feels odd to me.

**Comment by kmsquire** [comment]
> more readable and obvious than DefaultDict(x->identity) which feels odd to me.

That does seem odd—are you storing a function in the dictionary (certainly possible), or do you mean `DefaultDict(identity)`?

**Comment by mbauman** [comment]
Yes, I was storing a transformer function in a dict. Type stability is overrated. :)

**Comment by JeffBezanson** [comment]
This also has an advantage when the value might be expensive to compute, e.g. `Returns(expensive())` instead of `val = expensive(); x->val`. Admittedly that case is probably rare.

**Pull request merged**